compiler/modules/front/codeGen.bal

```diff
@@ -605,10 +628,10 @@ function codeGenMatchStmt(CodeGenContext cx, bir:BasicBlock startBlock, Environm
             if basis != () {
                 // Will need readOnlyIntersect when we have proper match patterns
                 t:SemType narrowedType = t:intersect(matchedType, clauseLooksLike[clauseIndex]);
-                clauseEnv = codeGenNarrowing(cx, stmtBlock, env, binding, narrowedType, basis, clause.opPos);
+                env.bindings = codeGenNarrowing(cx, stmtBlock, env, binding, narrowedType, basis, clause.opPos);
```

Comment:
Here you are mutating env, where the previous code was creating a new Env.  I doubt that is correct.